### ISSUE TITLE
Support databases other than sqlite

### ISSUE THREAD

**Issue by nathanloyer** [onc-healthit/inferno#511]
My team is looking at deploying inferno so we can use it to test various dev and production instances of our FHIR server. We want the test results to be persistent and for the inferno servers to be highly available, so relying on a database that is just storing data in a file in the local device storage is not going to be a viable solution. If the load balancer directs us to a different instance of inferno, trying to view the results will result in a 404. The test results also disappear when the service restarts.

In order to have a HA service, we need to be able to connect to a remote database. I've made some changes to support using postgres in my local copy of the code, and it is mostly working when just switching to the dm-postgres-adapter. However, I did have to make several updates to the data types in the data model classes because the strings that were being stored were not large enough. I can hack it together, but it would be nice to have support for this. I've probably missed various other data models that would have to be updated to support all of the test cases that my fhir server currently does not support.

As far as which databases to support, postgres is my personal preference, but any database that AWS RDS supports would be acceptable. Once one remote database is supported with the correct data types on the data models, it should be simple with the DataMapper framework to support several others.

**Comment by nathanloyer** [comment]
I just noticed that your development branch has removed the datamapper framework. My request remains the same, but I'm not sure how easy the new framework is to connect to different data sources.

**Comment by Jammjammjamm** [comment]
We replaced datamapper with ActiveRecord because datamapper is no longer maintained. Swapping out sqlite for postgres should require the following steps:

- Adding the `pg` gem to the `Gemfile`.
- Replacing `db/config.yml`.
- Making the schema changes you mentioned. You would create and run [migrations](https://guides.rubyonrails.org/active_record_migrations.html) to do this with ActiveRecord.

This is a skeleton `config.yml` file for postgres:
```
# PostgreSQL. Versions 9.3 and up are supported.
#
# Install the pg driver:
#   gem install pg
# On macOS with Homebrew:
#   gem install pg -- --with-pg-config=/usr/local/bin/pg_config
# On macOS with MacPorts:
#   gem install pg -- --with-pg-config=/opt/local/lib/postgresql84/bin/pg_config
# On Windows:
#   gem install pg
#       Choose the win32 build.
#       Install PostgreSQL and put its /bin directory on your path.
#
# Configure Using Gemfile
# gem 'pg'
#
default: &default
  adapter: postgresql
  encoding: unicode
  # For details on connection pooling, see Rails configuration guide
  # https://guides.rubyonrails.org/configuring.html#database-pooling
  pool: 5

development:
  <<: *default
  database: inferno_development

  # The specified database role being used to connect to postgres.
  # To create additional roles in postgres see `$ createuser --help`.
  # When left blank, postgres will use the default role. This is
  # the same name as the operating system user that initialized the database.
  #username: 

  # The password associated with the postgres role (username).
  #password:

  # Connect on a TCP socket. Omitted by default since the client uses a
  # domain socket that doesn't need configuration. Windows does not have
  # domain sockets, so uncomment these lines.
  #host: localhost

  # The TCP port the server listens on. Defaults to 5432.
  # If your server runs on a different port number, change accordingly.
  #port: 5432

  # Schema search path. The server defaults to $user,public
  #schema_search_path: myapp,sharedapp,public

  # Minimum log levels, in increasing order:
  #   debug5, debug4, debug3, debug2, debug1,
  #   log, notice, warning, error, fatal, and panic
  # Defaults to warning.
  #min_messages: notice

# Warning: The database defined as "test" will be erased and
# re-generated from your development database when you run "rake".
# Do not set this db to the same as development or production.
test:
  <<: *default
  database: inferno_test

# As with config/credentials.yml, you never want to store sensitive information,
# like your database password, in your source code. If your source code is
# ever seen by anyone, they now have access to your database.
#
# Instead, provide the password as a unix environment variable when you boot
# the app. Read https://guides.rubyonrails.org/configuring.html#configuring-a-database
# for a full rundown on how to provide these environment variables in a
# production deployment.
#
# On Heroku and other platform providers, you may have a full connection URL
# available as an environment variable. For example:
#
#   DATABASE_URL="postgres://myuser:mypass@localhost/somedatabase"
#
# You can use this database configuration with:
#
#   production:
#     url: <%= ENV['DATABASE_URL'] %>
#
production:
  <<: *default
  database: inferno_production
  username: inferno
  password: <%= ENV['INFERNO_DATABASE_PASSWORD'] %>

```

**Comment by nathanloyer** [comment]
Thanks, @Jammjammjamm. Are you part of the team that maintains Inferno? If so, is this a feature that you could add? Or is this something I would have to add if my team decides we need it? If my team does add support for it, is this something that should even be merged into this inferno repo?

This isn't particularly urgent for us, but we're going to be considering it within the next couple of months.

**Comment by Jammjammjamm** [comment]
Yes, I am. We are unlikely to add support for multiple databases. If your team decides to add support for this, I think that we would accept migrations to fix the schema, but probably not the rest. I noticed in the change to ActiveRecord that string lengths in the schema didn't seem right. I guess sqlite doesn't respect those lengths?

If you're planning to use docker, I think these changes could be done pretty easily with a custom docker image. Pseudocode dockerfile:
```
FROM inferno
COPY inferno's Gemfile with the 'sqlite3' replaced with 'pg'
RUN bundle install
COPY the db/config.yml
COPY the migrations
```

**Comment by nathanloyer** [comment]
I finished what I was working on at the time and proceeded to forget about this. Thanks for the response, if we decide to implement something I will look into creating a pull request here and we can discuss details at that point in time. 

Since sqlite just puts things in a local file I doubt it cares at all about enforcing string lengths. Actual DBMS's use string length to optimize storage so I can see why we'd only see the issues then. 

I suppose I will leave this issue open in case you decide later you want to implement this feature. Or you can close it if you want.

**Comment by nathanloyer** [comment]
@Jammjammjamm I recall hearing Inferno Program Edition now supports using Postgres. Is that feature planned to be patched into Community Edition?

We haven't tried hooking Program Edition up to Postgres yet, but we might try in the next couple of weeks.

**Comment by Jammjammjamm** [comment]
I don't expect that functionality to be moved over soon, as most of our non-program edition development is now focused on the next generation of inferno (https://github.com/inferno-community/inferno-core).

It should be pretty easy to stand up program edition with postgres based on the `docker-compose.postgres.yml` file in the program repo.

Since community already moved from datamapper to activerecord, changing the db shouldn't be too difficult. I think the main changes that would be needed are:
- add `pg` to the Gemfile
- update the schema to use larger fields (program schema could be used as a guide)
- update `db/config.yml` (program version could be used as a guide)
- add a postgres image to `docker-compose.yml` (program version could be used as a guide, or stand up postgres in whatever way makes sense for your environment)

**Comment by nathanloyer** [comment]
I wasn't aware of the inferno-core project, so thanks for showing that to me. I might try that out to resolve some of the other issues I've been noticing.

https://github.com/onc-healthit/inferno/issues/568
https://github.com/onc-healthit/inferno/issues/566

**Comment by Jammjammjamm** [comment]
This should be resolved by the 2.13.0 release. Please reopen this issue if it remains a problem.

**Comment by nathanloyer** [comment]
just want to note that things have been operating much more smoothly for me since starting to use the postgres DB over the summer
